### PR TITLE
Add Offline Support For Variety Sense (IOS ONLY)

### PR DESCRIPTION
**Pull request by zakikama**
Added the polar offline support its working but it takes a lot of time considering the amount of data encoded and decoded m still searching for a way to optimize this using protobuff instead of json or another way i would really appreciate it if u can do some searches from ur end feel free to contact me about anything ,
Sorry for the delay 😅

### CODE REVIEW

File: .idea/libraries/Dart_Packages.xml
Comment (Rexios80) [inline .idea/libraries/Dart_Packages.xml:1]: The `.idea` folder probably needs added to the `.gitignore` unless you created these by accident

File: ios/Classes/Codables.swift [hunk 1]
```diff
@@ -366,3 +374,249 @@ extension Date {
...
+        let type = PolarDeviceDataType.allCases[typeIndex]
+
+        data = PolarOfflineRecordingEntry(path: path, size: size, date: date, type: type)
+    }
+}
+
+
+
+
+class PolarEcgDataCodable: Encodable {
```
Comment (Rexios80) [inline ios/Classes/Codables.swift:420]: These types are already available with `PolarDataCodable<PolarEcgData>` etc

File: ios/Classes/SwiftPolarPlugin.swift [hunk 2]
```diff
@@ -112,6 +128,11 @@ public class SwiftPolarPlugin:
                 disableSdkMode(call, result)
             case "isSdkModeEnabled":
                 isSdkModeEnabled(call, result)
+            case "setOfflineRecordingTrigger":
```
Comment (Rexios80) [inline ios/Classes/SwiftPolarPlugin.swift:131]: Please group the new endpoints together at the end of the switch

File: ios/Classes/SwiftPolarPlugin.swift [hunk 3]
```diff
@@ -157,6 +180,199 @@ public class SwiftPolarPlugin:
...
+    //     }, onError: { error in
+    //         result(FlutterError(code: "Error starting offline offline recording", message: error.localizedDescription, details: nil))
+    //     })
+    // }
+
+
+
+
+
+func setOfflineRecordingTrigger(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
```
Comment (Rexios80) [inline ios/Classes/SwiftPolarPlugin.swift:209]: Please run [swift-format](https://github.com/swiftlang/swift-format) on all these changes

File: ios/Classes/SwiftPolarPlugin.swift [hunk 4]
```diff
@@ -157,6 +180,199 @@ public class SwiftPolarPlugin:
         result(nil)
     }
 
+
+
+    // func startOfflineRecording(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) throws {
```
Comment (Rexios80) [inline ios/Classes/SwiftPolarPlugin.swift:185]: Do not include commented out code in the final PR

File: ios/Classes/SwiftPolarPlugin.swift [hunk 5]
```diff
@@ -233,7 +467,7 @@ public class SwiftPolarPlugin:
                 }
                 exercises.append(data)
             }, onError: { error in
-                result(FlutterError(code: "Error listing exercises", message: error.localizedDescription, details: nil))
+                result(FlutterError(code: "Error listing exercises", message: error.localizedDescription, details: error.localizedDescription.description))
```
Comment (Rexios80) [inline ios/Classes/SwiftPolarPlugin.swift:470]: Isn't this a debug description of `error.localizedDescription` object? Is this useful?

File: lib/src/model/polar_acc_data.dart [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'polar_acc_data.g.dart';
+
+@JsonSerializable()
+class PolarAccData {
```
Comment (Rexios80) [inline lib/src/model/polar_acc_data.dart:6]: This type already exists in `polar_streaming.dart`

File: lib/src/model/polar_gyro_data.dart [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'polar_gyro_data.g.dart';
+
+@JsonSerializable()
+class PolarGyroData {
```
Comment (Rexios80) [inline lib/src/model/polar_gyro_data.dart:6]: This type already exists in `polar_streaming.dart`

File: lib/src/model/polar_magnetometer_data.dart [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'polar_magnetometer_data.g.dart';
+
+@JsonSerializable()
+class PolarMagnetometerData {
```
Comment (Rexios80) [inline lib/src/model/polar_magnetometer_data.dart:6]: This type already exists in `polar_streaming.dart`

File: lib/src/model/polar_offline_recording_entry.dart [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import 'package:json_annotation/json_annotation.dart';
+import 'package:polar/polar.dart';
+import 'package:polar/src/model/convert.dart';
+
+part 'polar_offline_recording_entry.g.dart';
+
+/// Polar Offline Recording Entry
+@JsonSerializable()
+class PolarOfflineRecordingEntry {
```
Comment (Rexios80) [inline lib/src/model/polar_offline_recording_entry.dart:9]: I would put this in the same file as `PolarOfflineRecordingData`

File: lib/src/model/polar_offline_recording_trigger.dart [hunk 10]
```diff
@@ -0,0 +1,37 @@
+import 'package:json_annotation/json_annotation.dart';
+import 'package:polar/src/model/polar_device_data_type.dart';
+import 'package:polar/src/model/polar_sensor_setting.dart';
+
+part 'polar_offline_recording_trigger.g.dart';
+
+@JsonSerializable(explicitToJson: true)
+class PolarOfflineRecordingTrigger {
```
Comment (Rexios80) [inline lib/src/model/polar_offline_recording_trigger.dart:8]: I would put this in the same file as `PolarOfflineRecordingData`

File: lib/src/model/polar_ppg_data.dart [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'polar_ppg_data.g.dart';
+
+@JsonSerializable()
+class PolarPpgData {
```
Comment (Rexios80) [inline lib/src/model/polar_ppg_data.dart:6]: This type already exists in `polar_streaming.dart`

File: lib/src/model/polar_ppi_data.dart [hunk 12]
```diff
@@ -0,0 +1,32 @@
+import 'package:json_annotation/json_annotation.dart';
+
+part 'polar_ppi_data.g.dart';
+
+@JsonSerializable()
+class PolarPpiData {
```
Comment (Rexios80) [inline lib/src/model/polar_ppi_data.dart:6]: This type already exists in `polar_streaming.dart`

File: lib/src/model/polar_streaming.g.dart [hunk 13]
```diff
@@ -1,6 +1,6 @@
 // GENERATED CODE - DO NOT MODIFY BY HAND
 
-// ignore_for_file: document_ignores, unnecessary_cast, require_trailing_commas
```
Comment (Rexios80) [inline lib/src/model/polar_streaming.g.dart:3]: Not sure how this got deleted. Merge with master and regenerate.

File: lib/src/polar_base.dart [hunk 14]
```diff
@@ -125,6 +126,18 @@ class Polar {
           ),
         );
         return;
+      case 'offlineRecordingTriggerSet':
+        try {
+          await setOfflineRecordingTrigger(
+            call.arguments[0],
+          );
+        } catch (e) {
+          developer.log(
```
Comment (Rexios80) [inline lib/src/polar_base.dart:135]: We should throw an exception here instead

File: polar.iml [hunk 15]
```diff
@@ -10,6 +10,7 @@
       <excludeFolder url="file://$MODULE_DIR$/build" />
       <excludeFolder url="file://$MODULE_DIR$/example/.pub" />
       <excludeFolder url="file://$MODULE_DIR$/example/build" />
+      <excludeFolder url="file://$MODULE_DIR$/example/.dart_tool" />
```
Comment (Rexios80) [inline polar.iml:13]: Is this file even supposed to be checked in?

**Review by Rexios80 — COMMENTED** [review]
This needs a lot of work to be mergable, but it's a good start. I'll work on adding CI to make our lives easier. We will need an Android implementation and tests as well.

**Comment by Rexios80** [comment]
> but it takes a lot of time considering the amount of data encoded and decoded

How slow is it? Are you sure the time isn't spent reading the data from the device rather than the transmission from native to dart?

**Comment by Rexios80** [comment]
Running `swift-format` on your code should resolve the conflict with `master`

**Comment by zakikama** [comment]
> > but it takes a lot of time considering the amount of data encoded and decoded
> 
> How slow is it? Are you sure the time isn't spent reading the data from the device rather than the transmission from native to dart?

i m not so sure for example a session of 58 min take about 5 min

**Comment by Rexios80** [comment]
What I'm asking is if that line is printed and then it takes 5 minutes to get the data in dart, or if it takes 5 minutes and then that line is printed

**Comment by zakikama** [comment]
it takes 5 minutes and then that line is printed

**Comment by Rexios80** [comment]
If that's the case, then no amount of work on our end will make it faster. Since we aren't actually doing any communication with dart until after that point it's not a serialization efficiency issue.

**Comment by Rexios80** [comment]
Maybe create an issue here? https://github.com/polarofficial/polar-ble-sdk

Even though bluetooth transmission speed is really slow I wouldn't think the data size would be large enough to be that slow.

**Comment by Rexios80** [comment]
If we assume the sensor has a memory size of 16MB and the recording takes up that entire amount, the maximum amount of transfer time would only be around 3 minutes. There must be something wrong here.

**Comment by Rexios80** [comment]
@zakikama are you planning on continuing work on this?

**Comment by abbjetmus** [comment]
Hi! I'm eager on getting this feature and can work on the android part of this.

**Comment by zakikama** [comment]
i managed to make it work nicely in IOS side of things i will push the new modifications today

**Comment by abbjetmus** [comment]
> i managed to make it work nicely in IOS side of things i will push the new modifications today

That sounds awesome, please do that and I will try to add the android part!

**Comment by abbjetmus** [comment]
@zakikama could you please update the PR so it can be included?

**Pull request closed without merging**